libbpf-tools/xfsslower.bpf.c

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#define BPF_NO_PRESERVE_ACCESS_INDEX
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "xfsslower.h"
+
+#define NULL    0
+
+const volatile pid_t targ_tgid = 0;
+const volatile __u64 min_lat = 0;
+
+struct piddata {
+	u64 ts;
+	loff_t start;
+	loff_t end;
+	struct file *fp;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u32);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline int
+probe_entry(struct file *fp, loff_t s, loff_t e)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddata.ts = bpf_ktime_get_ns();
+	piddata.start = s;
+	piddata.end = e;
+	piddata.fp = fp;
+	if (piddata.fp)
+		bpf_map_update_elem(&start, &pid, &piddata, 0);
```

Comment:
you can save a bunch of unnecessary work by exiting early `if (!fp)`, even before targ_tgid check. Also, piddata initialization won't be necessary, you are filling all the fields explicitly either way.

Reply:
thanks, I'll fix this.